File: app/src/main/java/io/mgba/Adapters/Interfaces/BaseAdapter.java
package io.mgba.Adapters.Interfaces;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;
import java.util.LinkedList;
import java.util.List;

public abstract class BaseAdapter<T> extends RecyclerView.Adapter<RecyclerView.ViewHolder> implements View.OnClickListener{
    private final RecyclerView instance;
    protected List<T> items = new LinkedList<>();
    protected Context mCtx;

    public BaseAdapter(Context context, RecyclerView recyclerView){
        this(new LinkedList<>(), recyclerView, context);
    }

    public BaseAdapter(List<T> items, RecyclerView recyclerView, Context context){
        this.items = items;
        this.instance = recyclerView;
        this.mCtx = context;
    }

    @Override
    public abstract RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType);

    protected int getPositionBasedOnView(View v){
        return instance.getChildAdapterPosition(v);
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    public void swap(List<T> items) {
        this.items = items;
        notifyDataSetChanged();
    }
}

File: app/src/main/java/io/mgba/adapters/interfaces/BaseAdapter.kt
package io.mgba.adapters.interfaces

import android.content.Context
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import org.lucasr.twowayview.listeners.SectionIndexer

abstract class BaseAdapter<T>(protected var items: List<T>, private val instance: RecyclerView, protected var context: Context) : RecyclerView.Adapter<RecyclerView.ViewHolder>(), View.OnClickListener, SectionIndexer {

    abstract override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder

    protected fun getPositionBasedOnView(v: View): Int {
        return instance.getChildAdapterPosition(v)
    }

    override fun getItemCount(): Int {
        return items.size
    }

    fun swap(items: List<T>) {
        this.items = items
        notifyDataSetChanged()
    }

    override fun getLetterForItem(position: Int): String {
        if(items.isNotEmpty() && items.size < position)
            return items.toString()
                        .toCharArray()[0]
                        .toUpperCase()
                        .toString()

        return "?"
    }

}